README.md

```diff
@@ -1,32 +1,128 @@
 
-The `SharedController` type is an extension of `Controller` that adds the ability to register multiple handlers on a
-shared controller instance.
+# Kubernetes Controllers
+_Note:_ This will be a very brief crash course. We encourage you to look up other resources regarding the topic, there are many! Two recommended resources are the [Sample Controller Project](https://github.com/kubernetes/sample-controller) and [Introduction To Kubernetes Controller Frameworks](https://github.com/aiyengar2/k8s-docs/blob/main/docs/controllers/03_frameworks.md).
 
-#### Simple Example
+From [the Kubernetes Documentation](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/)
+> In Kubernetes, a controller is a control loop that watches the shared state of the cluster through the apiserver and makes changes attempting to move the current state towards the desired state.
+
+The word "controller" get used more loosely than what the above definition might have you believe. In our own words: Controllers are loop handlers that act in response to Kubernetes resource events. In Lasso, we also use the word "controller" to refer to the entire structure that is responsible for registering this loop handler and providing it the resources it needs. In short, controllers are structures that bundle everything necessary to interact with and operate on Kubernetes types.
+
+In order to achieve the above goal, a few sub-goals need to be met:
+* the controller needs to know about resource type events
+* the controller needs to be passed relevant details regarding the event to a handler
+* (optional) the controller __may__ want to interact with Kubernetes resource types
+
+There are many frameworks to help achieve the above goals. Lasso is one of them. What most of these frameworks have in common, including lasso, is that they use functionality from client-go to achieve these goals. You could achieve these goals by using just client-go. For the remainder of this section, we will fly through how a controller is put together just using client-go concepts.
+
+An informer is used to achieve the above goals needed for a Kubernetes Controller. An informer possesses an event handler and a set of indexers. The underlying structs are different but a pseudo struct is shown below to illustrate this:
+
+```
+Informer
+	EventHandler
+	Indexers
+```
+
+The EventHandler can have different logic for create, update, and delete events. The event handler is where the controller comes in. You use it to configure the create, update, and delete event handlers.
+
+The indexers are a type of computational cache. They are used to map a key to an object, given a function, then the result is saved. Here it is easiest to think of them simply- they receive some metadata such as the name and namespace of the an object and return that object.
```

Comment:
>  I wouldn't want people to think each indexer has its own cache

Hmm, it is my understanding that `Indexer`s _are_ caches actually - indeed the `Indexer` interface extends `Store`. Now `Store`s (and, by extension, `Indexer`s) can in principle be used for anything - but in practice in client-go they are always used as caches of k8s resource objects (with various usage patterns and deletion policies).

One thing that confused me a lot is that `Indexer`s are not `Indexers`! `Indexers` (with the plural in the interface name!) is, indeed, a completely different concept - a map from `string` to `IndexFunc` - the latter being a function from `any` to a string slice.

Please see:
https://github.com/kubernetes/client-go/blob/1517ffb8d37c99e6a3a2842bcdee0aa271f0332b/tools/cache/index.go#L35-L36
vs
https://github.com/kubernetes/client-go/blob/master/tools/cache/index.go#L97-L98

Reply:
> Hmm, it is my understanding that `Indexer`s are caches actually

I agree `Indexer`s non-plural is essentially a cache. 
But the above struct was using plural `Indexers` which is not a cache.

``` 
Informer
	EventHandler
	Indexers
```
We are probably better off explaining that there are two concepts `Indexers`  and `Indexer` and distinguishing them separately to avoid conflation of the two. Especially because anyone using this package will most likely be working with the`Indexers` interface